pennylane/templates/subroutines/out_poly.py

```diff
@@ -0,0 +1,441 @@
+        ValueError: If the function is not defined with integer coefficients.
+
+    Example:
+        Given a polynomial function :math:`f(x, y) = x^2 + y`,
+        we can calculate :math:`f(3, 2)` as follows:
+
+        .. code-block:: python
+
+            wires = qml.registers({"x": 3, "y": 3, "output": 4})
+
+            def f(x, y):
+                return x ** 2 + y
+
+            @qml.qnode(qml.device("default.qubit", shots = 1))
+            def circuit():
+                # load values of x and y
+                qml.BasisEmbedding(3, wires=wires["x"])
+                qml.BasisEmbedding(2, wires=wires["y"])
+
+                # apply the polynomial
+                qml.OutPoly(
+                    f,
+                    input_registers = [wires["x"], wires["y"]],
+                    output_wires = wires["output"])
+
+                return qml.sample(wires=wires["output"])
+
+            print(circuit())
+
+        .. code-block:: pycon
+
+            >>> print(circuit())
+            [1 0 1 1]
+
+        The result, :math:`[1 0 1 1]`, is the binary representation of :math:`3^2 + 2 = 11`.
+        Note that the default value of `mod` in this example is :math:`2^{\text{len(output_wires)}} = 2^4 = 16`.
+        For more information on using `mod`, see the Usage Details section.
+
+
+    .. seealso:: :class:`~.PhaseAdder`
+
+    .. details::
+        :title: Usage Details
+
+        If the value of `mod` is not a power of two, then two auxiliary qubits must be provided.
+
+        .. code-block:: python
+
+            x_wires = [0, 1, 2]
+            y_wires = [3, 4, 5]
+            input_registers = [x_wires, y_wires]
+
+            output_wires = [6, 7, 8]
+            work_wires = [9,10]
+
+
+            def f(x, y):
+                return x ** 2 + y
+
+            @qml.qnode(qml.device("default.qubit", shots = 1))
+            def circuit():
+                # loading values for x and y
+                qml.BasisEmbedding(3, wires=x_wires)
+                qml.BasisEmbedding(2, wires=y_wires)
+
+                # applying the polynomial
+                qml.OutPoly(
+                    f,
+                    input_registers,
+                    output_wires,
+                    mod = 7,
+                    work_wires = work_wires
+                )
+
+                return qml.sample(wires=output_wires)
+
+        .. code-block:: pycon
+
+            >>> print(circuit())
+            [1 0 0]
+
+        The result, :math:`[1 0 0]`, is the binary representation of :math:`3^2 + 2  \; \text{modulo} \; 7 = 4`.
+        If the output wires are not initialized to zero, this value will be added to the solution. Generically,
+        the expression is definded as:
+
+        .. math::
+
+            \text{OutPoly}_{f, mod} |x_1 \rangle \dots |x_m \rangle |b \rangle
+            = |x_1 \rangle \dots |x_m \rangle |b + f(x_1, \dots, x_m) \mod mod \rangle.
```

Comment:
If you can include an example for this as well, it would be nice. Not required though.